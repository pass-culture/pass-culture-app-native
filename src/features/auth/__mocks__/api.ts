export const signIn = jest.fn()

export const useSignIn = jest.fn(() => signIn)

export const signUp = jest.fn()

export const useSignUp = jest.fn(() => signUp)

export const useDepositAmountsByAge = jest.fn().mockReturnValue({
  fifteenYearsOldDeposit: '20 €',
  sixteenYearsOldDeposit: '30 €',
  seventeenYearsOldDeposit: '30 €',
  eighteenYearsOldDeposit: '300 €',
})
