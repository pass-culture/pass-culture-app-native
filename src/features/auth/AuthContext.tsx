import NetInfo from '@react-native-community/netinfo'
import React, { memo, useCallback, useContext, useEffect, useMemo, useState } from 'react'
import { useQueryClient } from 'react-query'

import { api } from 'api/api'
import { refreshAccessToken } from 'api/apiHelpers'
import { SigninResponse } from 'api/gen'
import { useResetContexts } from 'features/auth/useResetContexts'
import { analytics, LoginRoutineMethod } from 'libs/analytics'
import { useAppStateChange } from 'libs/appState'
import { getAccessTokenStatus, getUserIdFromAccesstoken } from 'libs/jwt'
import { clearRefreshToken, saveRefreshToken } from 'libs/keychain'
import { eventMonitoring } from 'libs/monitoring'
import { QueryKeys } from 'libs/queryKeys'
import { BatchUser } from 'libs/react-native-batch'
import { storage } from 'libs/storage'

export interface IAuthContext {
  isLoggedIn: boolean
  setIsLoggedIn: (isLoggedIn: boolean) => void
}

const connectUserToBatchAndFirebase = (accessToken: string | null) => {
  if (!accessToken) return
  const userId = getUserIdFromAccesstoken(accessToken)
  if (userId) {
    BatchUser.editor().setIdentifier(userId.toString()).save()
    analytics.setUserId(userId)
    eventMonitoring.setUser({ id: userId.toString() })
  }
}

export const AuthContext = React.createContext<IAuthContext>({
  isLoggedIn: false,
  setIsLoggedIn: () => undefined,
})

export function useAuthContext(): IAuthContext {
  return useContext(AuthContext)
}

export const AuthWrapper = memo(function AuthWrapper({ children }: { children: JSX.Element }) {
  const [loading, setLoading] = useState(true)
  const [isLoggedIn, setIsLoggedIn] = useState(false)

  const readTokenAndConnectUser = useCallback(async () => {
    try {
      let accessToken = await storage.readString('access_token')
      const { isConnected } = await NetInfo.fetch()

      if (!!accessToken && !isConnected) {
        setIsLoggedIn(true)
        return
      }

      if (getAccessTokenStatus(accessToken) === 'expired') {
        // refreshAccessToken calls the backend to get a new access token
        // and also saves it to the storage
        accessToken = await refreshAccessToken(api)
      }

      if (getAccessTokenStatus(accessToken) === 'valid') {
        setIsLoggedIn(true)
        connectUserToBatchAndFirebase(accessToken)
      }
    } catch (err) {
      eventMonitoring.captureException(err)
      setIsLoggedIn(false)
    } finally {
      setLoading(false)
    }
  }, [])

  useEffect(() => {
    readTokenAndConnectUser()
  }, [readTokenAndConnectUser])

  useAppStateChange(readTokenAndConnectUser, () => void 0, [isLoggedIn])

  const value = useMemo(() => ({ isLoggedIn, setIsLoggedIn }), [isLoggedIn, setIsLoggedIn])

  if (loading) return null
  /**
   * warning: for a better data integrity, use setIsLoggedIn only from specific places
   * where it is needed:
   * - useLogoutRoutine: when logging out
   * - useLoginRoutine: when applying the logging in (signin or emailValidation)
   */
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
})

export function useLoginRoutine() {
  const { setIsLoggedIn } = useAuthContext()
  const resetContexts = useResetContexts()

  /**
   * Executes the minimal set of instructions required to proceed to the login
   * @param {SigninResponse} response
   * @param {LoginRoutineMethod} method The process that triggered the login routine
   */
  const loginRoutine = async (response: SigninResponse, method: LoginRoutineMethod) => {
    connectUserToBatchAndFirebase(response.accessToken)
    await saveRefreshToken(response.refreshToken)
    await storage.saveString('access_token', response.accessToken)
    analytics.logLogin({ method })
    setIsLoggedIn(true)
    resetContexts()
  }

  return loginRoutine
}

export function useLogoutRoutine(): () => Promise<void> {
  const queryClient = useQueryClient()
  const { setIsLoggedIn } = useAuthContext()

  return useCallback(async () => {
    try {
      BatchUser.editor().setIdentifier(null).save()
      analytics.logLogout()
      LoggedInQueryKeys.forEach((queryKey) => {
        queryClient.removeQueries(queryKey)
      })
      await storage.clear('access_token')
      await clearRefreshToken()
    } catch (err) {
      eventMonitoring.captureException(err)
    } finally {
      setIsLoggedIn(false)
    }
  }, [setIsLoggedIn])
}

// List of keys that are accessible only when logged in
// To clean when logging out
const LoggedInQueryKeys: QueryKeys[] = [
  QueryKeys.BOOKINGS,
  QueryKeys.FAVORITES,
  QueryKeys.FAVORITES_COUNT,
  QueryKeys.RECOMMENDATION_HITS,
  QueryKeys.RECOMMENDATION_OFFER_IDS,
  QueryKeys.REPORTED_OFFERS,
  QueryKeys.REPORT_OFFER_REASONS,
  QueryKeys.USER_PROFILE,
]
