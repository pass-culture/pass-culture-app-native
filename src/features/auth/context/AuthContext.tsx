import pick from 'lodash/pick'
import React, { memo, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react'
import { QueryObserverResult } from 'react-query'

import { api } from 'api/api'
import { UserProfileResponse } from 'api/gen'
import { useCookies } from 'features/cookies/helpers/useCookies'
import { navigateFromRef } from 'features/navigation/navigationRef'
// eslint-disable-next-line no-restricted-imports
import { amplitude } from 'libs/amplitude'
import { useAppStateChange } from 'libs/appState'
// eslint-disable-next-line no-restricted-imports
import { firebaseAnalytics } from 'libs/firebase/analytics'
import {
  computeTokenRemainingLifetimeInMs,
  getTokenStatus,
  getUserIdFromAccessToken,
} from 'libs/jwt'
import { getTokenExpirationDate } from 'libs/jwt/getTokenExpirationDate'
import { clearRefreshToken, getRefreshToken } from 'libs/keychain'
import { eventMonitoring } from 'libs/monitoring'
import { useNetInfoContext } from 'libs/network/NetInfoWrapper'
import { getAppVersion } from 'libs/packageJson'
import { QueryKeys } from 'libs/queryKeys'
import { BatchUser } from 'libs/react-native-batch'
import { usePersistQuery } from 'libs/react-query/usePersistQuery'
import { storage } from 'libs/storage'
import { getAge } from 'shared/user/getAge'

const MAX_AVERAGE_SESSION_DURATION_IN_MS = 60 * 60 * 1000
export interface IAuthContext {
  isLoggedIn: boolean
  setIsLoggedIn: (isLoggedIn: boolean) => void
  user?: UserProfileResponse
  refetchUser: () => Promise<QueryObserverResult<UserProfileResponse, unknown>>
  isUserLoading: boolean
}

const NAVIGATION_DELAY_FOR_EXPIRED_REFRESH_TOKEN_IN_MS = 1000

const navigateToLoginWithHelpMessage = () =>
  navigateFromRef('Login', { displayForcedLoginHelpMessage: true })

export const useConnectServicesRequiringUserId = (): ((accessToken: string | null) => void) => {
  const { setUserId: setUserIdToCookiesChoice } = useCookies()

  return useCallback(
    (accessToken) => {
      if (!accessToken) return

      const userId = getUserIdFromAccessToken(accessToken)
      if (userId) {
        BatchUser.editor().setIdentifier(userId.toString()).save()
        firebaseAnalytics.setUserId(userId)
        eventMonitoring.setUser({ id: userId.toString() })

        eventMonitoring.setExtras({
          accessTokenExpirationDate:
            getTokenExpirationDate(accessToken) ?? "can't get access token expiration date",
        })
        setUserIdToCookiesChoice(userId)
      }
    },
    [setUserIdToCookiesChoice]
  )
}

export const AuthContext = React.createContext<IAuthContext>({
  isLoggedIn: false,
  setIsLoggedIn: () => undefined,
  user: undefined,
  refetchUser: async () => ({} as QueryObserverResult<UserProfileResponse>),
  isUserLoading: false,
})

export function useAuthContext(): IAuthContext {
  return useContext(AuthContext)
}

export const AuthWrapper = memo(function AuthWrapper({
  children,
}: {
  children: React.JSX.Element
}) {
  const timeoutRef = useRef<NodeJS.Timeout>()
  const navigationTimeoutRef = useRef<NodeJS.Timeout>()
  const [loading, setLoading] = useState(true)
  const [isLoggedIn, setIsLoggedIn] = useState(false)
  const connectServicesRequiringUserId = useConnectServicesRequiringUserId()
  const {
    data: user,
    refetch: refetchUser,
    isLoading: isUserLoading,
  } = useUserProfileInfo(isLoggedIn)

  const readTokenAndConnectUser = useCallback(async () => {
    try {
      const refreshToken = await getRefreshToken()
      const accessToken = await storage.readString('access_token')
      const refreshTokenStatus = getTokenStatus(refreshToken)

      switch (refreshTokenStatus) {
        case 'unknown':
          setIsLoggedIn(false)
          return
        case 'expired':
          setIsLoggedIn(false)
          // We need to delay this navigation to avoid conflit between this navigation and the initial screen defined by react-navigation on app launch
          navigationTimeoutRef.current = setTimeout(async () => {
            navigateToLoginWithHelpMessage()
            await clearRefreshToken()
          }, NAVIGATION_DELAY_FOR_EXPIRED_REFRESH_TOKEN_IN_MS)
          return
        case 'valid':
          setIsLoggedIn(true)
          connectServicesRequiringUserId(accessToken)
          if (refreshToken) {
            eventMonitoring.setExtras({
              refreshTokenExpirationDate:
                getTokenExpirationDate(refreshToken) ?? "can't get refresh token expiration date",
            })
            const remainingLifetimeInMs = computeTokenRemainingLifetimeInMs(refreshToken)
            if (
              remainingLifetimeInMs &&
              remainingLifetimeInMs < MAX_AVERAGE_SESSION_DURATION_IN_MS
            ) {
              timeoutRef.current = globalThis.setTimeout(async () => {
                setIsLoggedIn(false)
                navigateToLoginWithHelpMessage()
                await clearRefreshToken()
              }, remainingLifetimeInMs)
            }
          }
          return
      }
    } catch (err) {
      eventMonitoring.captureException(err)
      setIsLoggedIn(false)
    } finally {
      setLoading(false)
    }
  }, [connectServicesRequiringUserId])

  useEffect(() => {
    readTokenAndConnectUser()

    return () => {
      if (timeoutRef.current) clearTimeout(timeoutRef.current)
      if (navigationTimeoutRef.current) clearTimeout(navigationTimeoutRef.current)
    }
  }, [readTokenAndConnectUser])

  useAppStateChange(readTokenAndConnectUser, () => void 0, [isLoggedIn])

  useEffect(() => {
    if (!user?.id) return

    const partialUser = pick(user, [
      'depositType',
      'eligibility',
      'eligibilityEndDatetime',
      'id',
      'isBeneficiary',
      'needsToFillCulturalSurvey',
    ])
    amplitude.setUserProperties({
      ...partialUser,
      ...(user.birthDate ? { age: getAge(user.birthDate) } : {}),
      status: user.status?.statusType, // eligible, beneficiaire, suspendu, etc
      appVersion: getAppVersion(),
    })
    amplitude.setUserId(user.id.toString())
  }, [user])

  const value = useMemo(
    () => ({ isLoggedIn, setIsLoggedIn, user, refetchUser, isUserLoading }),
    [isLoggedIn, setIsLoggedIn, user, refetchUser, isUserLoading]
  )

  if (loading) return null
  /**
   * warning: for a better data integrity, use setIsLoggedIn only from specific places
   * where it is needed:
   * - useLogoutRoutine: when logging out
   * - useLoginRoutine: when applying the logging in (signin or emailValidation)
   */
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
})

const STALE_TIME_USER_PROFILE = 5 * 60 * 1000
function useUserProfileInfo(isLoggedIn: boolean, options = {}) {
  const netInfo = useNetInfoContext()
  return usePersistQuery<UserProfileResponse>([QueryKeys.USER_PROFILE], () => api.getNativeV1Me(), {
    enabled: !!netInfo.isConnected && isLoggedIn,
    staleTime: STALE_TIME_USER_PROFILE,
    ...options,
  })
}
