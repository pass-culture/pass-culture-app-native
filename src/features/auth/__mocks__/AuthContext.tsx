import { useAuthContext as actualUseAuthContext } from '../AuthContext'

export const setIsLoggedIn = jest.fn()

export const useAuthContext = jest.fn().mockReturnValue({
  setIsLoggedIn,
  isLoggedIn: false,
}) as jest.MockedFunction<typeof actualUseAuthContext>

export const signOut = jest.fn()

export const useLogoutRoutine = jest.fn(() => signOut)

export const loginRoutine = jest.fn()

export const useLoginRoutine = jest.fn(() => loginRoutine)
