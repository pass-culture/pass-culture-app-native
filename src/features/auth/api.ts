import { api } from 'api/api'
import { AccountRequest, SigninRequest } from 'api/gen'
import { useLoginRoutine } from 'features/auth/AuthContext'

type SignInResponse =
  | {
      isSuccess: true
      content: undefined
    }
  | {
      isSuccess: false
      content?: {
        code: 'EMAIL_NOT_VALIDATED'
        general: string[]
      }
    }

export function useSignIn(): (data: SigninRequest) => Promise<SignInResponse> {
  const loginRoutine = useLoginRoutine()

  return async (body: SigninRequest) => {
    try {
      const response = await api.postnativev1signin(body, { credentials: 'omit' })
      if (!response) return { isSuccess: false }
      await loginRoutine(response, 'fromLogin')
      return { isSuccess: true }
    } catch (error) {
      return {
        isSuccess: false,
        content: error.statusCode
          ? error.content
          : { code: `OFFLINE`, general: [`Tu n'est pas connectÃ©`] },
      }
    }
  }
}

export function useSignUp(): (data: AccountRequest) => Promise<boolean> {
  return async (body: AccountRequest) => {
    try {
      const response = await api.postnativev1account(body, { credentials: 'omit' })
      return !!response
    } catch (error) {
      return false
    }
  }
}
