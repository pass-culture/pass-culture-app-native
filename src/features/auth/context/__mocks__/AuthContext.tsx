import { beneficiaryUser } from 'fixtures/user'

import { useAuthContext as actualUseAuthContext } from '../AuthContext'

export const setIsLoggedIn = jest.fn()

export const useAuthContext = jest.fn().mockReturnValue({
  setIsLoggedIn,
  isLoggedIn: true,
  user: beneficiaryUser,
  isUserLoading: false,
}) as jest.MockedFunction<typeof actualUseAuthContext>
