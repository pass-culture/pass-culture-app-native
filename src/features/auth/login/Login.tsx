import { t } from '@lingui/macro'
import { useNavigation } from '@react-navigation/native'
import React, { useRef, FunctionComponent, useState, useEffect } from 'react'
import { Keyboard, TouchableOpacity } from 'react-native'
import { useQuery } from 'react-query'
import styled from 'styled-components/native'

import { useSignIn } from 'features/auth/api'
import { useBackNavigation } from 'features/navigation/backNavigation'
import { NavigateToHomeWithoutModalOptions } from 'features/navigation/helpers'
import { UseNavigationType } from 'features/navigation/RootNavigator'
import { env } from 'libs/environment'
import { _ } from 'libs/i18n'
import { BottomContentPage } from 'ui/components/BottomContentPage'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { isValueEmpty } from 'ui/components/inputs/helpers'
import { InputError } from 'ui/components/inputs/InputError'
import { PasswordInput } from 'ui/components/inputs/PasswordInput'
import { TextInput } from 'ui/components/inputs/TextInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { getSpacing, Spacer, Typo } from 'ui/theme'

let INITIAL_IDENTIFIER = ''
let INITIAL_PASSWORD = ''

if (__DEV__) {
  INITIAL_IDENTIFIER = env.SIGNIN_IDENTIFIER
  INITIAL_PASSWORD = env.SIGNIN_PASSWORD
}

export const Login: FunctionComponent = function () {
  const isMounted = useRef(false)

  const [email, setEmail] = useState(INITIAL_IDENTIFIER)
  const [password, setPassword] = useState(INITIAL_PASSWORD)
  const [shouldShowErrorMessage, setShouldShowErrorMessage] = useState(false)
  const signIn = useSignIn()

  const shouldDisableLoginButton = isValueEmpty(email) || isValueEmpty(password)

  const { navigate } = useNavigation<UseNavigationType>()
  const complexGoBack = useBackNavigation<'Login'>()

  useEffect(() => {
    isMounted.current = true
    return () => {
      isMounted.current = false
    }
  }, [])

  const signInQuery = async () => {
    const signinResponse = await signIn({ identifier: email, password })
    if (signinResponse?.isSuccess) {
      navigate('Home', NavigateToHomeWithoutModalOptions)
    } else {
      const { code } = signinResponse?.content || {}
      if (code === 'EMAIL_NOT_VALIDATED') {
        navigate('SignupConfirmationEmailSent', { email })
      } else if (code === 'NETWORK_REQUEST_FAILED') {
        throw new Error('Offline')
      }
    }
    return signinResponse
  }

  const { refetch, isFetching } = useQuery('login', signInQuery, {
    cacheTime: 0,
    enabled: false,
  })

  async function handleSignin() {
    Keyboard.dismiss()
    setShouldShowErrorMessage(false)
    const { data } = await refetch()
    if (isMounted.current && !data?.isSuccess) {
      setShouldShowErrorMessage(true)
    }
  }

  function onClose() {
    navigate('Home', NavigateToHomeWithoutModalOptions)
  }

  function onForgottenPasswordClick() {
    navigate('ForgottenPassword')
  }

  return (
    <BottomContentPage>
      <ModalHeader
        title={_(t`Connecte-toi !`)}
        leftIcon={ArrowPrevious}
        onLeftIconPress={complexGoBack}
        rightIcon={Close}
        onRightIconPress={onClose}
      />
      <InputError
        visible={shouldShowErrorMessage}
        messageId="E-mail ou mot de passe incorrect"
        numberOfSpacesTop={5}
      />
      <Spacer.Column numberOfSpaces={7} />
      <StyledInput>
        <Typo.Body>{_(t`Adresse e-mail`)}</Typo.Body>
        <Spacer.Column numberOfSpaces={2} />
        <TextInput
          autoCapitalize="none"
          autoFocus={true}
          isError={shouldShowErrorMessage}
          keyboardType="email-address"
          onChangeText={setEmail}
          placeholder={_(/*i18n: email placeholder */ t`tonadresse@email.com`)}
          textContentType="emailAddress"
          value={email}
        />
      </StyledInput>
      <Spacer.Column numberOfSpaces={6} />
      <StyledInput>
        <Typo.Body>{_(t`Mot de passe`)}</Typo.Body>
        <Spacer.Column numberOfSpaces={2} />
        <PasswordInput
          value={password}
          onChangeText={setPassword}
          placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
          isError={shouldShowErrorMessage}
          textContentType="password"
        />
      </StyledInput>
      <Spacer.Column numberOfSpaces={7} />
      <ForgottenPasswordContainer>
        <TouchableOpacity onPress={onForgottenPasswordClick}>
          <Typo.ButtonText>{_(t`Mot de passe oubli√© ?`)}</Typo.ButtonText>
        </TouchableOpacity>
      </ForgottenPasswordContainer>
      <Spacer.Column numberOfSpaces={8} />
      <ButtonPrimary
        title={_(t`Se connecter`)}
        onPress={handleSignin}
        disabled={shouldDisableLoginButton || isFetching}
      />
    </BottomContentPage>
  )
}

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})

const ForgottenPasswordContainer = styled.View({
  flexDirection: 'row',
  justifyContent: 'flex-end',
  width: '100%',
  maxWidth: getSpacing(125),
})
