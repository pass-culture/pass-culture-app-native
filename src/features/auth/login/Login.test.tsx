import { act, fireEvent, render } from '@testing-library/react-native'
import { rest } from 'msw'
import React from 'react'
import waitForExpect from 'wait-for-expect'

import { navigate, useRoute } from '__mocks__/@react-navigation/native'
import { NavigateToHomeWithoutModalOptions, usePreviousRoute } from 'features/navigation/helpers'
import { env } from 'libs/environment'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { server } from 'tests/server'

import { useSignIn } from '../api'
import { AuthContext } from '../AuthContext'

import { Login } from './Login'

server.use(
  rest.post(env.API_BASE_URL + '/native/v1/signin', async (req, res, ctx) => res(ctx.status(401)))
)

jest.mock('features/navigation/helpers')
jest.mock('features/auth/api')

const mockUsePreviousRoute = usePreviousRoute as jest.Mock
const mockUseSignIn = useSignIn as jest.Mock

function renderLogin() {
  return render(
    reactQueryProviderHOC(
      <AuthContext.Provider value={{ isLoggedIn: true, setIsLoggedIn: jest.fn() }}>
        <Login />
      </AuthContext.Provider>
    )
  )
}

describe('<Login/>', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    mockUsePreviousRoute.mockReturnValue(null)
    useRoute.mockImplementation(() => ({
      params: {},
    }))
  })

  it('should redirect to home page WHEN signin is successful', async () => {
    mockUseSignIn.mockImplementationOnce(() => () => {
      return { isSuccess: true, content: undefined }
    })

    const { findByText } = renderLogin()

    const connexionButton = await findByText('Se connecter')
    await act(async () => {
      fireEvent.press(connexionButton)
    })

    await waitForExpect(() => {
      expect(navigate).toHaveBeenNthCalledWith(1, 'Home', NavigateToHomeWithoutModalOptions)
    })
  })

  it(
    'should redirect to SignupConfirmationEmailSent page' +
      ' WHEN signin has failed with EMAIL_NOT_VALIDATED code',
    async () => {
      mockUseSignIn.mockImplementationOnce(() => () => {
        return {
          isSuccess: false,
          content: {
            code: 'EMAIL_NOT_VALIDATED',
            general: ["L'email n'a pas été validé."],
          },
        }
      })

      const { findByText } = renderLogin()

      const connexionButton = await findByText('Se connecter')
      await act(async () => {
        fireEvent.press(connexionButton)
      })

      await waitForExpect(() => {
        expect(navigate).toHaveBeenNthCalledWith(1, 'SignupConfirmationEmailSent', {
          email: undefined,
        })
      })
    }
  )

  it('should show error message and error inputs AND not redirect to home page WHEN signin has failed', async () => {
    mockUseSignIn.mockImplementationOnce(() => () => {
      return { isSuccess: false, content: undefined }
    })

    const { findByText, toJSON } = renderLogin()
    const notErrorSnapshot = toJSON()

    const connexionButton = await findByText('Se connecter')
    await act(async () => {
      fireEvent.press(connexionButton)
    })

    await waitForExpect(() => {
      const errorSnapshot = toJSON()
      expect(notErrorSnapshot).toMatchDiffSnapshot(errorSnapshot)
      expect(navigate).not.toBeCalled()
    })
  })

  it('should enable login button when both text inputs are filled', async () => {
    const { getByPlaceholderText, toJSON } = renderLogin()
    const disabledButtonSnapshot = toJSON()

    const emailInput = getByPlaceholderText('tonadresse@email.com')
    const passwordInput = getByPlaceholderText('Ton mot de passe')
    fireEvent.changeText(emailInput, 'email@gmail.com')
    fireEvent.changeText(passwordInput, 'mypassword')

    await waitForExpect(() => {
      const enabledButtonSnapshot = toJSON()
      expect(disabledButtonSnapshot).toMatchDiffSnapshot(enabledButtonSnapshot)
    })
  })
})
