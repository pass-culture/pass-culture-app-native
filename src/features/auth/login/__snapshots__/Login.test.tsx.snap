// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login/> should enable login button when both text inputs are filled 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -559,10 +559,11 @@
                    },
                  ]
                }
                textContentType=\\"emailAddress\\"
                underlineColorAndroid=\\"transparent\\"
+               value=\\"email@gmail.com\\"
              />
            </View>
          </View>
          <View
            numberOfSpaces={6}
@@ -672,10 +673,11 @@
                    },
                  ]
                }
                textContentType=\\"password\\"
                underlineColorAndroid=\\"transparent\\"
+               value=\\"mypassword\\"
              />
              <View
                accessible={true}
                focusable={true}
                onClick={[Function onClick]}
@@ -794,22 +796,22 @@
              ]
            }
          />
          <View
            accessible={true}
-           focusable={false}
+           focusable={true}
            onClick={[Function onClick]}
            onResponderGrant={[Function onResponderGrant]}
            onResponderMove={[Function onResponderMove]}
            onResponderRelease={[Function onResponderRelease]}
            onResponderTerminate={[Function onResponderTerminate]}
            onResponderTerminationRequest={[Function onResponderTerminationRequest]}
            onStartShouldSetResponder={[Function onStartShouldSetResponder]}
            style={
              Object {
                \\"alignItems\\": \\"center\\",
-               \\"backgroundColor\\": \\"#ff99be\\",
+               \\"backgroundColor\\": \\"#eb0055\\",
                \\"borderRadius\\": 24,
                \\"borderWidth\\": 0,
                \\"flexDirection\\": \\"row\\",
                \\"height\\": 40,
                \\"justifyContent\\": \\"center\\","
`;

exports[`<Login/> should show error message and error inputs AND not redirect to home page WHEN signin has failed 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -454,10 +454,99 @@
                </RNSVGGroup>
              </RNSVGSvgView>
            </View>
          </View>
          <View
+           numberOfSpaces={5}
+           style={
+             Array [
+               Object {
+                 \\"height\\": 20,
+               },
+             ]
+           }
+           testID=\\"input-error-top-spacer\\"
+         />
+         <View
+           style={
+             Array [
+               Object {
+                 \\"alignItems\\": \\"center\\",
+                 \\"flexDirection\\": \\"row\\",
+                 \\"maxWidth\\": 500,
+               },
+             ]
+           }
+         >
+           <RNSVGSvgView
+             align=\\"xMidYMid\\"
+             bbHeight={24}
+             bbWidth={24}
+             focusable={false}
+             height={24}
+             meetOrSlice={0}
+             minX={0}
+             minY={0}
+             style={
+               Array [
+                 Object {
+                   \\"backgroundColor\\": \\"transparent\\",
+                   \\"borderWidth\\": 0,
+                 },
+                 Object {
+                   \\"flex\\": 0,
+                   \\"height\\": 24,
+                   \\"width\\": 24,
+                 },
+               ]
+             }
+             testID=\\"rule-icon-warn\\"
+             vbHeight={24}
+             vbWidth={24}
+             width={24}
+           >
+             <RNSVGGroup>
+               <RNSVGPath
+                 d=\\"M12 4.5a7.5 7.5 0 015.95 12.066.38.38 0 11-.604-.465 6.737 6.737 0 10-2.724 2.106.382.382 0 01.297.703A7.5 7.5 0 1112 4.5zm.118 3.779c.254 0 .463.143.496.33l.004.05v4.527c0 .21-.223.381-.5.381-.253 0-.462-.143-.495-.33l-.005-.051V8.66c0-.21.224-.381.5-.381zm.39 6.995a.508.508 0 10-1.015-.003.508.508 0 001.015.003\\"
+                 fill={4290117669}
+                 fillRule={0}
+                 propList={
+                   Array [
+                     \\"fill\\",
+                     \\"fillRule\\",
+                   ]
+                 }
+               />
+             </RNSVGGroup>
+           </RNSVGSvgView>
+           <View
+             style={
+               Array [
+                 Object {
+                   \\"paddingLeft\\": 8,
+                 },
+               ]
+             }
+           >
+             <Text
+               color=\\"#b60025\\"
+               style={
+                 Array [
+                   Object {
+                     \\"color\\": \\"#b60025\\",
+                     \\"fontFamily\\": \\"Montserrat-SemiBold\\",
+                     \\"fontSize\\": 12,
+                     \\"lineHeight\\": 16,
+                   },
+                 ]
+               }
+             >
+               E-mail ou mot de passe incorrect
+             </Text>
+           </View>
+         </View>
+         <View
            numberOfSpaces={7}
            style={
              Array [
                Object {
                  \\"height\\": 28,
@@ -501,18 +590,18 @@
                  },
                ]
              }
            />
            <View
-             isError={false}
+             isError={true}
              isFocus={false}
              style={
                Array [
                  Object {
                    \\"alignItems\\": \\"center\\",
                    \\"backgroundColor\\": \\"#ffffff\\",
-                   \\"borderColor\\": \\"#c7c7cc\\",
+                   \\"borderColor\\": \\"#b60025\\",
                    \\"borderRadius\\": 22,
                    \\"borderStyle\\": \\"solid\\",
                    \\"borderWidth\\": 1,
                    \\"flexDirection\\": \\"row\\",
                    \\"height\\": 40,
@@ -610,18 +699,18 @@
                  },
                ]
              }
            />
            <View
-             isError={false}
+             isError={true}
              isFocus={false}
              style={
                Array [
                  Object {
                    \\"alignItems\\": \\"center\\",
                    \\"backgroundColor\\": \\"#ffffff\\",
-                   \\"borderColor\\": \\"#c7c7cc\\",
+                   \\"borderColor\\": \\"#b60025\\",
                    \\"borderRadius\\": 22,
                    \\"borderStyle\\": \\"solid\\",
                    \\"borderWidth\\": 1,
                    \\"flexDirection\\": \\"row\\",
                    \\"height\\": 40,"
`;
